### PR TITLE
Implement local type inference for `let` statements

### LINKED ISSUE #161: Local type inference
For instance, in many cases `let`-expressions don't need type ascriptions.

### PR DESCRIPTION
**Pull request by Gusarich**
resolves #161

### CODE REVIEW

File: src/grammar/grammar.ohm [hunk 1]
```diff
@@ -98,7 +98,8 @@ Tact {
               | StatementRepeat
               | StatementUntil
     StatementBlock = "{" Statement* "}"
-    StatementLet = let id ":" Type "=" Expression ";"
+    StatementLet = let id ":" Type "=" Expression ";" --withType
+                 | let id "=" Expression ";" --withoutType
```
Comment (anton-trunov) [inline src/grammar/grammar.ohm]: How about we just make the type ascription optional?
```suggestion
    StatementLet = let id (":" Type)? "=" Expression ";"
```

**Comment by anton-trunov** [comment]
Let's fix the merge conflict in CHANGELOG and move the tests to their new location

**Comment by Gusarich** [comment]
@anton-trunov done

File: src/test/e2e-emulated/contracts/local-type-inference.tact
Comment (anton-trunov) [inline src/test/e2e-emulated/contracts/local-type-inference.tact:1]: We should cover more primitive types in tests:
- `Slice`,
- `Cell`,
- `Builder`,
- `String`,
- `StringBuilder`,
- `StateInit`

and also the compound types:
- maps,
- structs,
- optionals

File: src/grammar/clone.ts [hunk 2]
```diff
@@ -28,7 +28,6 @@ export function cloneNode<T extends ASTNode>(src: T): T {
     } else if (src.kind === "statement_let") {
         return cloneASTNode({
             ...src,
-            type: cloneASTNode(src.type),
```
Comment (anton-trunov) [inline src/grammar/clone.ts:31]: you are sure we won't run into issues in the future if we just ignore the type here?

for instance, `def_argument` has the following:
```
    } else if (src.kind === "def_argument") {
        return cloneASTNode({
            ...src,
            type: cloneASTNode(src.type),
        });
```

I'd say we should either clone all types or not clone any.

**Review by anton-trunov — CHANGES_REQUESTED** [review]
Look good! Let's add some more tests (see my comment below) and there is one concern with AST cloning

**Comment by anton-trunov** [comment]
There is a case when local type inference should break: `let foo = null`. It's not clear what type to assign to `foo` in this case, it can be any optional type, e.g. `Int?` or any map type, like `map<Int, Int>` or `map<Address, Cell>`, etc. We should add negative tests for things like that.

File: src/grammar/grammar.ts
Comment (anton-trunov) [inline src/grammar/grammar.ts:1]: let's fix the `StatementLet` parameter names too:
```
StatementLet(_letKwd, id, _optColon, optType, _equals, expression, _semicolon) {
```

File: src/grammar/clone.ts (hunk 2 above)
Comment (Gusarich) [reply src/grammar/clone.ts:31]: those aren't cloned in `ASTField`, `ASTConstant`, `ASTFunction` and `ASTNativeFunction` too. Should I add that type cloning to all of these?

File: src/grammar/grammar.ts
Comment (Gusarich) [reply src/grammar/grammar.ts:1]: What's wrong with these?

Comment (anton-trunov) [reply src/grammar/grammar.ts:1]: Those are currently named `_colon` and `type`, but this PR makes them optional, and everywhere else in `grammar.ts` optional items are prefixed with `opt`

File: src/grammar/clone.ts (hunk 2 above)
Comment (anton-trunov) [reply src/grammar/clone.ts:31]: good point, let's clone them all

**Comment by Gusarich** [comment]
@anton-trunov everything's done!

**Review by anton-trunov — APPROVED** [review]
Let's go!

**Pull request merged**